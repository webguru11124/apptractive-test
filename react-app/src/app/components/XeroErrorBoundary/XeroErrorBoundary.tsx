import { Button } from '@mui/material';
import { ErrorBoundary } from 'react-error-boundary';
import { useTranslation } from 'react-i18next';
interface XeroErrorBoundaryProps {
  children: React.ReactNode;
}
export function XeroErrorBoundary({ children }: XeroErrorBoundaryProps) {
  const { t } = useTranslation();
  return (
    <ErrorBoundary
      fallbackRender={({ resetErrorBoundary }) => (
        <div>
          {t('xeroError', { ns: 'xero' })}
          <Button
            onClick={() => {
              resetErrorBoundary();
            }}
          >
            {t('tryAgain', { ns: 'common' })}
          </Button>
        </div>
      )}
    >
      {children}
    </ErrorBoundary>
  );
}
