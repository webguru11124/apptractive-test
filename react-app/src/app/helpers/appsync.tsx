import { AuthOptions } from 'aws-appsync-auth-link';
import cdkOutput from '../output.json';
import { Auth } from 'aws-amplify';
import {
  ApolloClient,
  ApolloLink,
  gql,
  HttpLink,
  InMemoryCache
} from '@apollo/client';
import { createAuthLink } from 'aws-appsync-auth-link';
import { createSubscriptionHandshakeLink } from 'aws-appsync-subscription-link';
import {
  isLoggedInVar,
  subInVar,
  typeDefs
} from '../graphql';

/**
 * Configure AppSync client
 */
const httpLink = new HttpLink({
  uri: cdkOutput.FULAppSyncAPIStack.GraphQLAPIURL
});

export const configureAppSyncClient = () => {
  const url = cdkOutput.FULAppSyncAPIStack.GraphQLAPIURL; // TODO: duplicated url?
  const region = 'us-east-1';

  const auth: AuthOptions = {
    type: 'AMAZON_COGNITO_USER_POOLS',
    jwtToken: async () =>
      (await Auth.currentSession()).getIdToken().getJwtToken()
    //credentials: async () => credentials, // Required when using IAM-based auth.
  };

  const link = ApolloLink.from([
    createAuthLink({ url, region, auth }),
    createSubscriptionHandshakeLink({ url, region, auth }, httpLink)
  ]);

  return new ApolloClient({
    link,
    cache: new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            isLoggedIn: {
              read() {
                return isLoggedInVar();
              }
            },
            sub: {
              read() {
                return subInVar();
              }
            }
          }
        }
      }
    }),
    connectToDevTools: true,
    typeDefs: gql(typeDefs)
  });
};
