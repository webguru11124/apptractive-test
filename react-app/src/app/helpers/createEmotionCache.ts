import createCache from '@emotion/cache';
export type { EmotionCache } from '@emotion/cache';

export function createEmotionCache() {
  return createCache({ key: 'css' });
}
