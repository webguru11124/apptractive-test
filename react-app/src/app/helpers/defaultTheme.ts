import { PaletteMode, ThemeOptions } from '@mui/material';

export const defaultTheme = (mode: PaletteMode): ThemeOptions => ({
  components: {
    MuiCssBaseline: {
      styleOverrides: {
        html: {
          height: '100vh',
        },
        body: {
          height: '100vh',
        },
        '#root': {
          height: '100vh',
        },
      },
    },
  },
});
