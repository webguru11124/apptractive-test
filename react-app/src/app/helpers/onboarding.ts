import { OnboardingStatus, User } from '../graphql';

/**
 * Gets the next step of onboarding based on user type
 * @param user
 */
export const getNextOnboardingStep = (user: User) => {
  const { onboardingStatus } = user;
  switch (onboardingStatus) {
    case OnboardingStatus.NAME:
      return OnboardingStatus.PROFILE_IMAGE;
    case OnboardingStatus.PROFILE_IMAGE:
      return OnboardingStatus.INTERESTS;
    case OnboardingStatus.INTERESTS:
      return OnboardingStatus.PLANS;
    case OnboardingStatus.PLANS:
      return OnboardingStatus.COMPLETED;
    default:
      return OnboardingStatus.COMPLETED;
  }
};
