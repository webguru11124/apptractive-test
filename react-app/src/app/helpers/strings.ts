export function convertColumnNamesToDisplayText(inputString: string) {
  const newStr = inputString
    .replace(/([A-Z])/g, ' $1') // insert a space before all capital letters
    .replace(/^./, (str: string) => str.toUpperCase()) // capitalize the first letter of each word
    .replace(/([a-z0-9])([A-Z])/g, '$1 $2'); // insert a space between last lowercase & uppercase letter
  return newStr;
}
