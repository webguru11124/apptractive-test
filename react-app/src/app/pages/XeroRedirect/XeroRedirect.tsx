import { gql, useMutation } from '@apollo/client';
import { OperationVariables } from '@apollo/client/core/types';
import {
  xeroCreateTokenSet as XERO_CREATE_TOKEN_SET,
  XeroScopeSet,
} from '../../graphql';
import { Typography } from '@mui/material';
import { Auth } from 'aws-amplify';
import { useEffect } from 'react';
import { useTranslation } from 'react-i18next';
import { useSearchParams } from 'react-router-dom';

/* eslint-disable-next-line */
export interface XeroRedirectProps {}

export function XeroRedirect(props: XeroRedirectProps) {
  const { t } = useTranslation();
  const [searchParams] = useSearchParams();
  const [xeroCreateTokenSet] = useMutation(gql(XERO_CREATE_TOKEN_SET), {});

  const errorCode = searchParams.get('error');

  useEffect(() => {
    const getAuthUser = async () => {
      try {
        const authUser = await Auth.currentAuthenticatedUser();
        console.log('authUser: ', authUser);
      } catch (err) {
        console.log('ERROR Auth.currentAuthenticatedUser: ', err);
      }
    };

    getAuthUser();
  }, []);

  useEffect(() => {
    const createTokenSet = async () => {
      const url = window.location.href;
      const options: OperationVariables = {
        variables: {
          input: {
            url,
            scopeSet: XeroScopeSet.ACCOUNTING,
          },
        },
      }

      try {
        const { data } = await xeroCreateTokenSet(options);

        console.log('data: ', data);
      } catch (err) {
        console.log('ERROR create xero token set', err);
      }
    };

    createTokenSet();
  }, [xeroCreateTokenSet]);

  return (
    <>
      <Typography>Xero redirected</Typography>
      {errorCode && (
        <Typography color="error">
          {t('xeroError', { ns: 'xero' })} ({errorCode})
        </Typography>
      )}
    </>
  );
}

export default XeroRedirect;
