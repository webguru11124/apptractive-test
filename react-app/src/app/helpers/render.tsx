import { MockedProvider } from '@apollo/client/testing';
import { CacheProvider } from '@emotion/react';
import { CssBaseline } from '@mui/material';
import { ThemeProvider } from '@mui/material/styles';
import { ThemeProvider as EmotionThemeProvider } from '@emotion/react';
import { render } from '@testing-library/react';
import { PropsWithChildren, ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { DesignSystemContextProvider } from '../components';
import { getTheme } from '../constants/theme';
import { createEmotionCache } from './createEmotionCache';
import { defaultTheme } from './defaultTheme';

const defaultCache = createEmotionCache();
const theme = getTheme(defaultTheme('light'));

const AllTheProviders = ({ children }: PropsWithChildren) => {
  return (
    <MockedProvider>
      <BrowserRouter>
        <CacheProvider value={defaultCache}>
          <EmotionThemeProvider theme={theme}>
            <ThemeProvider theme={theme}>
              <CssBaseline />
              <DesignSystemContextProvider theme={theme}>
                  {children}
              </DesignSystemContextProvider>
            </ThemeProvider>
          </EmotionThemeProvider>
        </CacheProvider>
      </BrowserRouter>
    </MockedProvider>
  );
};

const customRender = (ui: ReactElement, options = {}) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
