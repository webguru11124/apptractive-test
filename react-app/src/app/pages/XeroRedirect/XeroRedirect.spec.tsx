import { MockedProvider, wait } from '@apollo/client/testing';
import { render } from '../../helpers/render';
import {
  xeroCreateTokenSet,
  XeroScopeSet,
} from '../../graphql';
import XeroRedirect, { XeroRedirectProps } from './XeroRedirect';
import { gql } from '@apollo/client';
import { ErrorBoundary } from 'react-error-boundary';
import {  act, screen, waitFor } from '@testing-library/react'
import { XeroErrorBoundary } from '../../components/XeroErrorBoundary/XeroErrorBoundary';

const XERO_CREATE_TOKEN_SET = gql`${xeroCreateTokenSet}`;
const request = {
  query: XERO_CREATE_TOKEN_SET,
  variables: {
    input: {
      url:"https://localhost:4200/xero-redirect",
      scopeSet: [XeroScopeSet.ACCOUNTING],
    }
  }
}
const response = {
  token:"123-456",
  expireIn:12345,
  user:{
    email:"abc@abc.com",
    givenName:"abc",
    familyName:"def",
    __typename:"user"
  },
  __typename:"xeroCreateTokenSet"
  
}
const mocks = [
  {
    request,
    result:{data:{xeroCreateTokenSet:{...response}}}
  },
  {
    delay: 1000 ,
    request,
    result:{data:{xeroCreateTokenSet:{...response}}}
  },
  {
    request,
    error: new Error("An error occurred")
  },
];

const renderComponent = (props?:XeroRedirectProps)=>render(  
  <XeroErrorBoundary>
    <MockedProvider mocks={mocks} addTypename={false}>
      <XeroRedirect {...props}/>
    </MockedProvider>
  </XeroErrorBoundary>)

describe('XeroRedirect', () => {
  it('should render successfully', async () => {
    await act(async ()=>{
      const { baseElement } = renderComponent();
      await waitFor(() => {
        expect(baseElement).toBeTruthy();
      });
    })
  });

  it('should render correctly',async () => {
    await  act(async ()=>{
      const { baseElement } = renderComponent();
      await waitFor(() => {
        expect(baseElement).toMatchSnapshot();
      });
    })
    
  });
});
