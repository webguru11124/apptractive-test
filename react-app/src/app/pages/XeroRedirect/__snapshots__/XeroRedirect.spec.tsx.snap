// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`XeroRedirect > should render correctly 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-gatsgs"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
      >
        xeroRedirection
      </h3>
      <a
        class=""
        href="/dashboard"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1etjjuw-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          goToDashboard
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </a>
    </div>
  </div>
</body>
`;

exports[`XeroRedirect > should render correctly 2`] = `<body />`;

exports[`XeroRedirect > should render correctly 3`] = `
<body>
  <div />
</body>
`;

exports[`XeroRedirect > should render successfully 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-gatsgs"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
      >
        xeroRedirection
      </h3>
      <a
        class=""
        href="/dashboard"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1etjjuw-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          goToDashboard
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </a>
    </div>
  </div>
</body>
`;
